File: packages/golang/validator.go
// Package validator provides JSON Schema validation for Schema Resume documents.
package validator

import (
	_ "embed"
	"encoding/json"
	"fmt"

	"github.com/xeipuuv/gojsonschema"
)

//go:embed schemas/schema.json
var schemaJSON []byte

//go:embed schemas/meta-schema.json
var metaSchemaJSON []byte

//go:embed schemas/context.jsonld
var contextJSON []byte

// Validator provides resume validation functionality
type Validator struct {
	schema     *gojsonschema.Schema
	schemaData map[string]interface{}
	metaSchema map[string]interface{}
	context    map[string]interface{}
}

// ValidationResult contains the validation outcome
type ValidationResult struct {
	Valid  bool            `json:"valid"`
	Errors []ValidationError `json:"errors"`
}

// ValidationError represents a single validation error
type ValidationError struct {
	Field       string `json:"field"`
	Type        string `json:"type"`
	Description string `json:"description"`
	Value       string `json:"value,omitempty"`
}

// NewValidator creates a new resume validator instance
func NewValidator() (*Validator, error) {
	// Parse schema data
	var schemaData map[string]interface{}
	if err := json.Unmarshal(schemaJSON, &schemaData); err != nil {
		return nil, fmt.Errorf("failed to parse schema: %w", err)
	}

	// Parse meta-schema
	var metaSchema map[string]interface{}
	if err := json.Unmarshal(metaSchemaJSON, &metaSchema); err != nil {
		return nil, fmt.Errorf("failed to parse meta-schema: %w", err)
	}

	// Parse context
	var context map[string]interface{}
	if err := json.Unmarshal(contextJSON, &context); err != nil {
		return nil, fmt.Errorf("failed to parse context: %w", err)
	}

	// Create schema loader
	schemaLoader := gojsonschema.NewBytesLoader(schemaJSON)
	schema, err := gojsonschema.NewSchema(schemaLoader)
	if err != nil {
		return nil, fmt.Errorf("failed to create schema: %w", err)
	}

	return &Validator{
		schema:     schema,
		schemaData: schemaData,
		metaSchema: metaSchema,
		context:    context,
	}, nil
}

// Validate validates a resume document
func (v *Validator) Validate(resume interface{}) (*ValidationResult, error) {
	documentLoader := gojsonschema.NewGoLoader(resume)
	result, err := v.schema.Validate(documentLoader)
	if err != nil {
		return nil, fmt.Errorf("validation failed: %w", err)
	}

	validationResult := &ValidationResult{
		Valid:  result.Valid(),
		Errors: make([]ValidationError, 0),
	}

	if !result.Valid() {
		for _, err := range result.Errors() {
			validationResult.Errors = append(validationResult.Errors, ValidationError{
				Field:       err.Field(),
				Type:        err.Type(),
				Description: err.Description(),
				Value:       fmt.Sprintf("%v", err.Value()),
			})
		}
	}

	return validationResult, nil
}

// ValidateJSON validates a resume from JSON bytes
func (v *Validator) ValidateJSON(resumeJSON []byte) (*ValidationResult, error) {
	var resume interface{}
	if err := json.Unmarshal(resumeJSON, &resume); err != nil {
		return nil, fmt.Errorf("failed to parse resume JSON: %w", err)
	}
	return v.Validate(resume)
}

// GetSchema returns the JSON Schema as a map
func (v *Validator) GetSchema() map[string]interface{} {
	return v.schemaData
}

// GetMetaSchema returns the meta-schema as a map
func (v *Validator) GetMetaSchema() map[string]interface{} {
	return v.metaSchema
}

// GetContext returns the JSON-LD context as a map
func (v *Validator) GetContext() map[string]interface{} {
	return v.context
}

// ValidateResume is a convenience function to validate a resume
func ValidateResume(resume interface{}) (*ValidationResult, error) {
	validator, err := NewValidator()
	if err != nil {
		return nil, err
	}
	return validator.Validate(resume)
}

// ValidateResumeJSON is a convenience function to validate resume from JSON bytes
func ValidateResumeJSON(resumeJSON []byte) (*ValidationResult, error) {
	validator, err := NewValidator()
	if err != nil {
		return nil, err
	}
	return validator.ValidateJSON(resumeJSON)
}

File: packages/php/src/Validator.php
<?php

namespace SchemaResume;

use JsonSchema\Validator as JsonValidator;
use JsonSchema\Constraints\Constraint;

/**
 * Validator for Schema Resume JSON documents.
 */
class Validator
{
    private object $schema;
    private object $metaSchema;
    private object $context;
    private string $schemasDir;

    /**
     * Create a new Validator instance.
     *
     * @param string|null $schemaPath Optional path to custom schema file
     * @throws \RuntimeException If schema files cannot be loaded
     */
    public function __construct(?string $schemaPath = null)
    {
        $this->schemasDir = __DIR__ . '/../schemas';
        
        if ($schemaPath !== null) {
            $this->schema = $this->loadJson($schemaPath);
        } else {
            $this->schema = $this->loadJson($this->schemasDir . '/schema.json');
        }
        
        $this->metaSchema = $this->loadJson($this->schemasDir . '/meta-schema.json');
        $this->context = $this->loadJson($this->schemasDir . '/context.jsonld');
    }

    /**
     * Load JSON file.
     *
     * @param string $path Path to JSON file
     * @return object Decoded JSON object
     * @throws \RuntimeException If file cannot be loaded or parsed
     */
    private function loadJson(string $path): object
    {
        if (!file_exists($path)) {
            throw new \RuntimeException("Schema file not found: {$path}");
        }
        
        $content = file_get_contents($path);
        if ($content === false) {
            throw new \RuntimeException("Failed to read schema file: {$path}");
        }
        
        $json = json_decode($content);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \RuntimeException("Invalid JSON in schema file: {$path}");
        }
        
        return $json;
    }

    /**
     * Validate a resume document.
     *
     * @param array|object|string $resume Resume data as array, object, or JSON string
     * @return array Validation result with 'valid' and 'errors' keys
     * @throws \InvalidArgumentException If resume data is invalid
     */
    public function validate(array|object|string $resume): array
    {
        $resumeData = $this->parseResume($resume);
        
        $validator = new JsonValidator();
        $validator->validate($resumeData, $this->schema, Constraint::CHECK_MODE_NORMAL);
        
        $errors = [];
        if (!$validator->isValid()) {
            foreach ($validator->getErrors() as $error) {
                $errors[] = [
                    'path' => $error['property'],
                    'message' => $error['message'],
                    'constraint' => $error['constraint'] ?? null,
                ];
            }
        }
        
        return [
            'valid' => $validator->isValid(),
            'errors' => $errors,
        ];
    }

    /**
     * Parse resume data from various formats.
     *
     * @param array|object|string $resume Resume data
     * @return object Parsed resume object
     * @throws \InvalidArgumentException If resume data is invalid
     */
    private function parseResume(array|object|string $resume): object
    {
        if (is_string($resume)) {
            $decoded = json_decode($resume);
            if (json_last_error() !== JSON_ERROR_NONE) {
                throw new \InvalidArgumentException('Invalid JSON: ' . json_last_error_msg());
            }
            return $decoded;
        }
        
        if (is_array($resume)) {
            return (object) $resume;
        }
        
        return $resume;
    }

    /**
     * Get the JSON Schema.
     *
     * @return object The schema
     */
    public function getSchema(): object
    {
        return $this->schema;
    }

    /**
     * Get the meta-schema.
     *
     * @return object The meta-schema
     */
    public function getMetaSchema(): object
    {
        return $this->metaSchema;
    }

    /**
     * Get the JSON-LD context.
     *
     * @return object The context
     */
    public function getContext(): object
    {
        return $this->context;
    }
}
